crates/ruff_db/src/lib.rs

```diff
@@ -0,0 +1,75 @@
+use std::hash::BuildHasherDefault;
+
+use rustc_hash::FxHasher;
+use salsa::DbWithJar;
+
+use crate::vfs::VfsFile;
+
+pub mod vfs;
+
+pub(crate) type FxDashMap<K, V> = dashmap::DashMap<K, V, BuildHasherDefault<FxHasher>>;
+
+#[salsa::jar(db=Db)]
+pub struct Jar(VfsFile);
+
+pub trait Db: DbWithJar<Jar> {
+    /// Interns a file path and returns a salsa `File` ingredient.
+    ///
+    /// The operation is guaranteed to always succeed, even if the path doesn't exist, isn't accessible, or if the path points to a directory.
+    /// In these cases, a file with status [`FileStatus::Deleted`] is returned.
+    fn file(&self, path: &camino::Utf8Path) -> VfsFile;
+
+    /// Interns a path to a vendored file and returns a salsa `File` ingredient.
+    fn vendored_file(&self, path: &camino::Utf8Path) -> Option<VfsFile>;
```

Comment:
@AlexWaygood I think we ultimately want two different entry functions for regular files, and vendored files. Mainly because I think it's a bit annoying if you're working with a vendored path and you then need to convert it to a `VfsPath` just to get the `VfsFile`. Funnily, the first thing that the implementation would do is to dispatch on the path type.